File: src/api/tutor.py
"""The ``/tutor`` resource: importing a spreadsheet and taking it as a quiz.

Every endpoint is gated behind the ``tutor`` feature toggle -- off, the
resource answers 404 the same as a route that does not exist, so the plugin
stays invisible until it is turned on in the settings panel.
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, File, Form, HTTPException, Request, UploadFile, status
from pydantic import ValidationError

from connectors.hollow_connector import NoteIndex

from .dependencies import Hollow
from .dependencies_settings import require_feature
from .dependencies_tutor import TutorStore
from .schemas_tutor import (
    TutorAttemptCreate,
    TutorCommit,
    TutorImportOut,
    TutorNoteCreate,
    TutorNoteUpdate,
    TutorQuizOut,
    TutorQuizRename,
    TutorQuizSummaryOut,
    TutorReferenceFolder,
)

router = APIRouter(
    prefix="/tutor",
    tags=["tutor"],
    dependencies=[Depends(require_feature("tutor"))],
)


def _resolve_reference(index: NoteIndex, label: str, folder: str = "") -> str | None:
    """Resolve a reference label to a hollow-relative path.

    Args:
        index: The hollow's note index.
        label: The reference label to resolve -- a note's name or path.
        folder: A hollow-relative folder to restrict the search to, subfolders
            included. Empty searches the whole hollow.

    Returns:
        The path of the first indexed note the label names within scope, or
        None when nothing matches.
    """
    wanted = label.strip().lower()
    if not wanted:
        return None
    stem = wanted[: -len(".md")] if wanted.endswith(".md") else wanted
    matches = [
        note for note in index.notes if note.name.lower() == stem or note.path.lower() == wanted
    ]
    if folder:
        matches = [
            note for note in matches if note.folder == folder or note.folder.startswith(f"{folder}/")
        ]
    return matches[0].path if matches else None


@router.post(
    "/imports",
    response_model=TutorImportOut,
    status_code=status.HTTP_201_CREATED,
    summary="Stage a .csv or .xlsx spreadsheet for column mapping",
)
async def stage_import(
    request: Request,
    tutor: TutorStore,
    file: UploadFile = File(description="A .csv or .xlsx spreadsheet."),
) -> TutorImportOut:
    """Parse an uploaded spreadsheet and hold it in memory for column mapping.

    Args:
        request: The request, read for the configured size limit.
        tutor: The tutor connector.
        file: The uploaded file, sent as ``multipart/form-data``.

    Returns:
        Its header, a short preview of its rows, and the id the mapping step
        commits against.

    Raises:
        HTTPException: The upload is empty, unnamed, or larger than the
            configured limit.
    """
    content = await file.read()
    if not content:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "The uploaded file is empty")

    limit = request.app.state.max_import_bytes
    if len(content) > limit:
        raise HTTPException(
            status.HTTP_413_REQUEST_ENTITY_TOO_LARGE,
            f"The upload is larger than the {limit} bytes this hollow accepts",
        )

    if not file.filename:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "The uploaded file has no name")

    table = tutor.stage_import(file.filename, content)
    return TutorImportOut(
        import_id=table.id,
        filename=table.filename,
        columns=list(table.columns),
        preview_rows=[list(row) for row in table.rows[:5]],
    )


@router.post(
    "/imports/{import_id}/commit",
    response_model=TutorQuizOut,
    status_code=status.HTTP_201_CREATED,
    summary="Map columns and build a quiz set out of a staged import",
)
async def commit_import(
    import_id: str,
    request: Request,
    tutor: TutorStore,
    hollow: Hollow,
    payload: str = Form(description="The column mapping, as JSON -- see TutorCommit."),
    images: list[UploadFile] = File(
        default=[], description="Images to match to questions, when include_images is set."
    ),
) -> TutorQuizOut:
    """Turn a staged import into a persisted quiz set.

    A reference label is resolved against the hollow's note index the same
    way a wiki-link is, so it matches a note by name regardless of which
    folder holds it. An uploaded image is matched to every row whose
    ``image_columns`` cells name it, and stored under the quiz set once it
    is built.

    Args:
        import_id: The staged import to build from.
        request: The request, read for the configured size limit.
        tutor: The tutor connector.
        hollow: The hollow connector, read for its note index.
        payload: The column mapping and the name to give the quiz set, sent
            as a form field carrying JSON since images ride along as files.
        images: The images to match against ``image_columns``, sent as
            ``multipart/form-data``.

    Returns:
        The quiz set that was built and persisted.

    Raises:
        HTTPException: ``payload`` is not valid JSON for a commit, or an
            image is larger than the configured limit.
    """
    try:
        commit = TutorCommit.model_validate_json(payload)
    except ValidationError as error:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, str(error)) from error

    note_index = hollow.index()

    def resolve_reference(label: str) -> str | None:
        return _resolve_reference(note_index, label)

    image_uploads: list[tuple[str, bytes]] = []
    if commit.include_images:
        limit = request.app.state.max_image_bytes
        for image in images:
            content = await image.read()
            if not image.filename or not content:
                continue
            if len(content) > limit:
                raise HTTPException(
                    status.HTTP_413_REQUEST_ENTITY_TOO_LARGE,
                    f"'{image.filename}' is larger than the {limit} bytes this hollow accepts",
                )
            image_uploads.append((image.filename, content))

    quiz = tutor.commit_import(
        import_id,
        name=commit.name,
        question_column=commit.question_column,
        answer_columns=tuple(commit.answer_columns),
        correct_column=commit.correct_column,
        reference_column=commit.reference_column,
        reference_separator=commit.reference_separator.strip() if commit.multi_reference else "",
        resolve_reference=resolve_reference,
        image_columns=tuple(commit.image_columns) if commit.include_images else (),
        image_uploads=tuple(image_uploads),
    )
    return TutorQuizOut.model_validate(quiz)


@router.get("/quizzes", response_model=list[TutorQuizSummaryOut], summary="List every quiz set")
def list_quizzes(tutor: TutorStore) -> list[TutorQuizSummaryOut]:
    """Every persisted quiz set, without its questions."""
    return [TutorQuizSummaryOut.model_validate(quiz) for quiz in tutor.list_quizzes()]


@router.get("/quizzes/{quiz_id}", response_model=TutorQuizOut, summary="Read one quiz set")
def get_quiz(quiz_id: str, tutor: TutorStore) -> TutorQuizOut:
    """One persisted quiz set, questions included."""
    return TutorQuizOut.model_validate(tutor.get_quiz(quiz_id))


@router.patch("/quizzes/{quiz_id}", response_model=TutorQuizOut, summary="Rename a quiz set")
def rename_quiz(quiz_id: str, payload: TutorQuizRename, tutor: TutorStore) -> TutorQuizOut:
    """Give a quiz set a new name."""
    return TutorQuizOut.model_validate(tutor.rename_quiz(quiz_id, payload.name))


@router.patch(
    "/quizzes/{quiz_id}/reference-folder",
    response_model=TutorQuizOut,
    summary="Search one folder of the hollow for a quiz's references",
)
def set_reference_folder(
    quiz_id: str,
    payload: TutorReferenceFolder,
    tutor: TutorStore,
    hollow: Hollow,
) -> TutorQuizOut:
    """Re-resolve every question's reference within one folder of the hollow.

    Each question keeps the reference label it was imported with; only the
    path it resolves to is refreshed. Useful when the hollow holds more than
    one note of the same name and the wrong one won on the whole-hollow search.

    Args:
        quiz_id: The quiz set to update.
        payload: The folder to search, subfolders included; empty for the
            whole hollow.
        tutor: The tutor connector.
        hollow: The hollow connector, read for its note index.

    Returns:
        The quiz set, its references refreshed.
    """
    note_index = hollow.index()
    folder = payload.folder.strip().strip("/")

    def resolve_reference(label: str) -> str | None:
        return _resolve_reference(note_index, label, folder)

    quiz = tutor.set_reference_folder(quiz_id, folder, resolve_reference)
    return TutorQuizOut.model_validate(quiz)


@router.post(
    "/quizzes/{quiz_id}/attempts",
    response_model=TutorQuizOut,
    status_code=status.HTTP_201_CREATED,
    summary="Record a completed attempt at a quiz set",
)
def record_attempt(quiz_id: str, payload: TutorAttemptCreate, tutor: TutorStore) -> TutorQuizOut:
    """Record how one run through the quiz went, for its statistics."""
    quiz = tutor.record_attempt(
        quiz_id,
        mode=payload.mode,
        question_count=payload.question_count,
        correct_count=payload.correct_count,
    )
    return TutorQuizOut.model_validate(quiz)


@router.delete(
    "/quizzes/{quiz_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete a quiz set",
)
def delete_quiz(quiz_id: str, tutor: TutorStore) -> None:
    """Delete a quiz set."""
    tutor.delete_quiz(quiz_id)


@router.post(
    "/quizzes/{quiz_id}/questions/{question_id}/notes",
    response_model=TutorQuizOut,
    status_code=status.HTTP_201_CREATED,
    summary="Add a note to a question, after it has been answered",
)
def add_note(
    quiz_id: str, question_id: str, payload: TutorNoteCreate, tutor: TutorStore
) -> TutorQuizOut:
    """Add a note to a question."""
    return TutorQuizOut.model_validate(tutor.add_note(quiz_id, question_id, payload.text))


@router.patch(
    "/quizzes/{quiz_id}/questions/{question_id}/notes/{note_id}",
    response_model=TutorQuizOut,
    summary="Change one note's text",
)
def update_note(
    quiz_id: str, question_id: str, note_id: str, payload: TutorNoteUpdate, tutor: TutorStore
) -> TutorQuizOut:
    """Change one note's text, independently of the question's other notes."""
    return TutorQuizOut.model_validate(tutor.update_note(quiz_id, question_id, note_id, payload.text))


@router.delete(
    "/quizzes/{quiz_id}/questions/{question_id}/notes/{note_id}",
    response_model=TutorQuizOut,
    summary="Delete one note",
)
def delete_note(quiz_id: str, question_id: str, note_id: str, tutor: TutorStore) -> TutorQuizOut:
    """Delete one note, independently of the question's other notes."""
    return TutorQuizOut.model_validate(tutor.delete_note(quiz_id, question_id, note_id))

File: src/frontend/src/api/tutor.ts
/** The tutor plugin's own types and HTTP calls: importing a spreadsheet and taking it as a quiz. */

import { json, request } from './http'

/** A spreadsheet staged for column mapping, before it becomes a quiz. */
export interface TutorImport {
  /** The id this staged import is held under; the mapping step commits against it. */
  importId: string
  /** The name the upload arrived with. */
  filename: string
  /** The header row. */
  columns: string[]
  /** The first few data rows, shown while columns are picked. */
  previewRows: string[][]
}

/** The body that turns a staged import into a quiz set. */
export interface TutorCommitPayload {
  /** The name to give the quiz set. */
  name: string
  /** The column holding each question's text. */
  questionColumn: string
  /** The columns holding each answer, in display order. */
  answerColumns: string[]
  /** The column naming the correct answer of each row. */
  correctColumn: string
  /** The column holding a reference note's name, if any. */
  referenceColumn: string | null
  /** Whether the reference column may name more than one source, joined by `referenceSeparator`. */
  multiReference: boolean
  /** The literal that splits a reference label into several source names, used only when
   *  `multiReference` is set. */
  referenceSeparator: string
  /** Whether uploaded images should be matched to the rows that name them. */
  includeImages: boolean
  /** The columns to search for an uploaded image's name, used only when `includeImages`
   *  is set. Usually the question column, but answer columns may be included too. */
  imageColumns: string[]
}

/** One note left on a question, after it was answered. */
export interface TutorNote {
  /** The note's id, stable within its question. */
  id: string
  /** The note's text. */
  text: string
  /** When it was written, ISO 8601. */
  createdAt: string
}

/** One question of a quiz set. */
export interface TutorQuestion {
  /** The question's id, stable within its quiz set. */
  id: string
  /** The question text. */
  question: string
  /** The possible answers, in the order they show. */
  answers: string[]
  /** The indices into `answers` that are correct -- more than one for a multi-answer question. */
  correctIndices: number[]
  /** The reference label, empty when the row had none. */
  reference: string
  /** The hollow-relative paths each source in `reference` resolved to; a source that
   *  matched nothing is left out. */
  referencePaths: string[]
  /** The notes left on this question, oldest first. */
  notes: TutorNote[]
  /** The tutor-relative paths of the images matched to this question at import time. */
  imagePaths: string[]
}

/** How a quiz's questions are chosen for one attempt. */
export type TutorMode = 'sequential' | 'random' | 'manual'

/** One completed run through a quiz set, kept for its statistics. */
export interface TutorAttempt {
  /** The attempt's id. */
  id: string
  /** How its questions were chosen. */
  mode: TutorMode
  /** How many questions the attempt covered. */
  questionCount: number
  /** How many of them were answered correctly. */
  correctCount: number
  /** When the attempt was recorded, ISO 8601. */
  completedAt: string
}

/** A quiz set, questions included. */
export interface TutorQuiz {
  /** The quiz set's id. */
  id: string
  /** The name it was given at import time. */
  name: string
  /** When it was built, ISO 8601. */
  createdAt: string
  /** Its questions, in source order. */
  questions: TutorQuestion[]
  /** Every completed run through it, oldest first. */
  attempts: TutorAttempt[]
  /** The hollow-relative folder its references were last resolved against, subfolders
   *  included; empty means the whole hollow. */
  referenceFolder: string
  /** The literal that splits a reference label into several source names, empty when
   *  each label names a single source. */
  referenceSeparator: string
}

/** A quiz set without its questions, for a list view. */
export interface TutorQuizSummary {
  /** The quiz set's id. */
  id: string
  /** The name it was given at import time. */
  name: string
  /** When it was built, ISO 8601. */
  createdAt: string
  /** How many questions it holds. */
  questionCount: number
}

export const tutorApi = {
  /** Stage a .csv or .xlsx spreadsheet, parsed and ready for column mapping. */
  stageTutorImport(file: File): Promise<TutorImport> {
    const form = new FormData()
    form.append('file', file)
    return request<TutorImport>('/tutor/imports', { method: 'POST', body: form })
  },

  /** Map columns of a staged import and build a persisted quiz set out of it, images
   *  it should be matched against rows included. */
  commitTutorImport(
    importId: string,
    payload: TutorCommitPayload,
    images: File[] = [],
  ): Promise<TutorQuiz> {
    const form = new FormData()
    form.append('payload', JSON.stringify(payload))
    for (const image of images) form.append('images', image)
    return request<TutorQuiz>(`/tutor/imports/${encodeURIComponent(importId)}/commit`, {
      method: 'POST',
      body: form,
    })
  },

  /** Every persisted quiz set, without its questions. */
  listTutorQuizzes(): Promise<TutorQuizSummary[]> {
    return request<TutorQuizSummary[]>('/tutor/quizzes')
  },

  /** One persisted quiz set, questions included. */
  getTutorQuiz(id: string): Promise<TutorQuiz> {
    return request<TutorQuiz>(`/tutor/quizzes/${encodeURIComponent(id)}`)
  },

  /** Give a quiz set a new name. */
  renameTutorQuiz(id: string, name: string): Promise<TutorQuiz> {
    return request<TutorQuiz>(`/tutor/quizzes/${encodeURIComponent(id)}`, json('PATCH', { name }))
  },

  /** Re-resolve a quiz set's references within one folder of the hollow, subfolders included. */
  setTutorReferenceFolder(id: string, folder: string): Promise<TutorQuiz> {
    return request<TutorQuiz>(
      `/tutor/quizzes/${encodeURIComponent(id)}/reference-folder`,
      json('PATCH', { folder }),
    )
  },

  /** Record how one completed run through a quiz set went, for its statistics. */
  recordTutorAttempt(
    id: string,
    mode: TutorMode,
    questionCount: number,
    correctCount: number,
  ): Promise<TutorQuiz> {
    return request<TutorQuiz>(
      `/tutor/quizzes/${encodeURIComponent(id)}/attempts`,
      json('POST', { mode, questionCount, correctCount }),
    )
  },

  /** Delete a quiz set. */
  deleteTutorQuiz(id: string): Promise<void> {
    return request<void>(`/tutor/quizzes/${encodeURIComponent(id)}`, { method: 'DELETE' })
  },

  /** Add a note to a question, after it has been answered. */
  addTutorNote(quizId: string, questionId: string, text: string): Promise<TutorQuiz> {
    return request<TutorQuiz>(
      `/tutor/quizzes/${encodeURIComponent(quizId)}/questions/${encodeURIComponent(questionId)}/notes`,
      json('POST', { text }),
    )
  },

  /** Change one note's text, independently of the question's other notes. */
  updateTutorNote(
    quizId: string,
    questionId: string,
    noteId: string,
    text: string,
  ): Promise<TutorQuiz> {
    return request<TutorQuiz>(
      `/tutor/quizzes/${encodeURIComponent(quizId)}/questions/${encodeURIComponent(questionId)}` +
        `/notes/${encodeURIComponent(noteId)}`,
      json('PATCH', { text }),
    )
  },

  /** Delete one note, independently of the question's other notes. */
  deleteTutorNote(quizId: string, questionId: string, noteId: string): Promise<TutorQuiz> {
    return request<TutorQuiz>(
      `/tutor/quizzes/${encodeURIComponent(quizId)}/questions/${encodeURIComponent(questionId)}` +
        `/notes/${encodeURIComponent(noteId)}`,
      { method: 'DELETE' },
    )
  },
}
